discretize/CylMesh.py

```diff
@@ -123,72 +227,531 @@ def vectorCCx(self):
+        )
+
+
+    @property
+    def areaFz(self):
+        if getattr(self, '_areaFz', None) is None:
+            areaFzFull = self._areaFzFull
             if self.isSymmetric:
-                self._edge = 2*pi*self.gridN[:, 0]
+                self._areaFz = areaFzFull
             else:
-                raise NotImplementedError('edges not yet implemented for 3D '
-                                          'cyl mesh')
-        return self._edge
+                self._areaFz = (
+                    self._deflationMatrix('Fz', withHanging=False) *
+                    areaFzFull
+                )
+        return self._areaFz
+
+    @property
+    def _areaFull(self):
+        return np.r_[self._areaFxFull, self._areaFyFull, self._areaFzFull]
 
     @property
     def area(self):
         """Face areas"""
-        if getattr(self, '_area', None) is None:
-            if self.nCy > 1:
-                raise NotImplementedError('area not yet implemented for 3D '
-                                          'cyl mesh')
-            areaR = np.kron(self.hz, 2*pi*self.vectorNx)
-            areaZ = np.kron(np.ones_like(self.vectorNz), pi*(self.vectorNx**2 -
-                            np.r_[0, self.vectorNx[:-1]]**2))
-            self._area = np.r_[areaR, areaZ]
-        return self._area
+        # if getattr(self, '_area', None) is None:
+        if self.isSymmetric is True:
+            return np.r_[self.areaFx, self.areaFz]
+        else:
+            return np.r_[self.areaFx, self.areaFy, self.areaFz]
 
     @property
     def vol(self):
         """Volume of each cell"""
         if getattr(self, '_vol', None) is None:
-            if self.nCy > 1:
-                raise NotImplementedError('vol not yet implemented for 3D '
-                                          'cyl mesh')
-            az = pi*(self.vectorNx**2 - np.r_[0, self.vectorNx[:-1]]**2)
-            self._vol = np.kron(self.hz, az)
+            if self.isSymmetric:
+                az = pi*(self.vectorNx**2 - np.r_[0, self.vectorNx[:-1]]**2)
+                self._vol = np.kron(self.hz, az)
+            else:
+                self._vol = np.kron(
+                    self.hz, np.kron(
+                        self.hy,
+                        0.5 * (self.vectorNx[1:]**2 - self.vectorNx[:-1]**2)
+                    )
+                )
         return self._vol
 
     ####################################################
+    # Active and Hanging Edges and Faces
+    ####################################################
+
+    @property
+    def _ishangingFx(self):
+        if getattr(self, '__ishangingFx', None) is None:
+            hang_x = np.zeros(self._ntNx, dtype=bool)
+            hang_x[0] = True
+            self.__ishangingFx = np.kron(
+                np.ones(self.nCz, dtype=bool),
+                np.kron(
+                    np.ones(self.nCy, dtype=bool),
+                    hang_x
+                )
+            )
+        return self.__ishangingFx
```

Comment:
Double underscores are not ideal to use. Not an issue here, but good to stay aware of.


> Private name mangling: When an identifier that textually occurs in a class definition begins with two or more underscore characters and does not end in two or more underscores, it is considered a private name of that class. Private names are transformed to a longer form before code is generated for them. The transformation inserts the class name in front of the name, with leading underscores removed, and a single underscore inserted in front of the class name. For example, the identifier __spam occurring in a class named Ham will be transformed to _Ham__spam. This transformation is independent of the syntactical context in which the identifier is used. If the transformed name is extremely long (longer than 255 characters), implementation defined truncation may happen. If the class name consists only of underscores, no transformation is done.

https://docs.python.org/3.5/reference/expressions.html?highlight=mangling#index-5